docs/RFCS/20190619_constraint_conformance_report.md

```diff
@@ -0,0 +1,516 @@
+no information on over-replicated ranges - although hopefully those will go away
+once we get atomic group membership changes. Also, the way in which these
+counters are computed is defective: the counters are incoherent and, for
+example, if all the nodes holding a partition go away, the ranges in that
+partition are not counted (see
+https://github.com/cockroachdb/cockroach/issues/19644).
+
+Besides the general need for an administrator to know that everything is
+copacetic, there are some specific cases where the lack of information is
+particularly unfortunate:
+
+
+1. Testing fault tolerance: A scenario people ran into was testing CRDB’s
+resiliency to networking failures. A cluster is configured with 3 Availability
+Zones (AZs) and the data is constrained to require a replica in every AZ (or
+even without constraints, the implicit diversity requirement would drive towards
+the same effect). One would think that an AZ can go away at any time and the
+cluster should survive. But that’s only colloquially true. After an AZ goes away
+and subsequently recovers, another AZ cannot disappear right away without
+availability loss; the 2nd AZ has to stay up until constraint conformance is
+re-established. Otherwise, the 2nd AZ likely has 2 replicas for the ranges that
+were migrated away during the first outage.
+As things stand, it’s entirely unclear how one is supposed to know when it’s
+safe to bring down the 2nd AZ. (Of course, here we’re talking about disaster
+testing, not the use of node decomissioning.)
+1. Partitioning a table and running performance tests that assume the partioning
+is in place. For example, say I want to partition my data between Romania and
+Bulgaria and test that my Bulgarian clients indeed start getting local
+latencies. Well, partitioning is obviously not instantaneous and so it’s
+entirely unclear how long I’m supposed to wait until I can conduct my test.
+Same with lease preferences.
+
+## Proposal
+
+We’ll be introducing three ways for admins to observe different aspects of
+replication state: a cluster-wide constraint conformance report visible in the
+AdminUI, new jobs for tracking the progress of constraint changes, and two
+crdb_internal virtual tables that can be queried for similar information to the
+report - one table would list zone-level constraint violations, another would
+list the violations on a more granular per-object basis.
+For providing some insights into the diversity of ranges in different zone
+configs, there's also going to be a crdb_internal virtual table that lists the
+localities whose unavailability would result in range unavailability (e.g. "if
+locality US goes away, half of the ranges in the default zone would lose
+quorum").
+
+
+### crdb_internal.replication_report(cached bool (default true))
+
+We'll introduce the `crdb_internal.replication_report()` function, which
+returns a report about constraint violations and critical localities as a
+(single) JSON record.
```

Comment:
I don't see the need to bundle all that information into a single JSON record. That would make it harder to add more information when needed. The normal way of exposing information is via tables. We can fit this information into tables. For example:
	
### zone_violations

 zone_id | type       | text             | updated_at          | time_since_status_changed | failed_ranges | failed_mb | failed_ranges_ps | failed_mb_ps
---------|------------|------------------|---------------------|---------------------------|---------------|-----------|------------------|--------------
234 | constraint | {+country=us:1}  | 2019-08-07 18:04:27 |                  00:01:22 |           12  |       768 |            -0.01 |        -0.64 
234 | constraint | {+ssd}           | 2019-08-07 18:04:28 |    5 mons 3 days 00:00:02 |            0  |         0 |                0 |            0 
234 | lease      | [+region=na]     | 2019-08-07 18:04:28 |    1 mons 2 days 00:20:09 |            0  |         0 |                0 |            0 

### localities

locality           | updated_at          | ranges_at_risk  | mb_at_risk  | ranges_at_risk_ps | mb_at_risk_ps 
-------------------|---------------------|-----------------|-------------|-------------------|---------------
region=na,dc=east  | 2019-08-07 18:04:27 |               0 |           0 |                 0 |             0 |
region=na,dc=west  | 2019-08-07 18:04:27 |               5 |         320 |             -0.01 |         -0.64 |
region=na          | 2019-08-07 18:04:28 |             320 |       20480 |             -0.13 |         -8.32 |


### zone_ranges

zone_id | total_ranges | under_replicated | over_replicated | unavailable |
--------|--------------|------------------|-----------------|-------------
0 |         2520 |               21 |               9 |           0 |
234 |          320 |                0 |               2 |           0 |

Reply:
> I don't see the need to bundle all that information into a single JSON record. That would make it harder to add more information when needed. 

Why harder? Every schemaless bro will tell you that evolving the schema of a JSON record is easier than that of a SQL table, not harder.

What are the `updated_at` and `time_since_status_changed` fields exactly? How would the `updated_at` field end up different between rows?

For the `localities` table, don't you think we want to also report the risk at the zone level?
And did you ignore the "node id as last component of locality" part on purpose, or just for brevity? (I.e. do you agree with reporting on critical nodes?)

Where do you stand on the `system` vs `crdb_internal` namespace question and how it relates to storing the velocity info vs computing it on demand?